### PR TITLE
Refactor - version 2

### PR DESCRIPTION
**Pull request by florian-huber**
- Create parent classes in ``BaseSimilarity.py``
- Implemented in ``CosineGreedy``, ``CosineHungarian``, ``FingerprintSimilarity``, ``IntersectMZ``, ``ModifiedCosine``, ``ParentmassMatch`` and ``Scores``
- removed ``calculate_scores_parallel()`` and the ``xyzParallel`` similarity scores
- label ``Scores.calculate()`` as deprecated.

### CODE REVIEW

**Comment by florian-huber** [comment]
Sorry to send this big PR your way @sverhoeven. Here I implemented similarity functions that inherit from ``BaseSimilarityFunction`` and that only return numpy arrays as scores (so no ``Scores`` objects). But I would be definitely open to discuss alternatives!

**Review by sverhoeven — CHANGES_REQUESTED** [review]
That is a lot of changes. Nice work to testing and documenting everything and keep the code coverage up.

* see suggestions in inlined conversations
* Missing is a Change and Deprecated section in the CHANGELOG 
* Having a class ending with `Function` makes it a bit awkward to talk about it. Could we rename `BaseSimilarityFunction` to `BaseSimilarity` or `BaseSpectrumSimilarity`?

File: matchms/similarity/FingerprintSimilarity.py [hunk 1]
```diff
@@ -1,20 +1,64 @@
...
 import numpy
+from matchms.similarity.vector_similarity_functions import cosine_similarity
 from matchms.similarity.vector_similarity_functions import \
     cosine_similarity_matrix
+from matchms.similarity.vector_similarity_functions import dice_similarity
 from matchms.similarity.vector_similarity_functions import \
     dice_similarity_matrix
+from matchms.similarity.vector_similarity_functions import jaccard_index
 from matchms.similarity.vector_similarity_functions import \
     jaccard_similarity_matrix
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply matchms/similarity/FingerprintSimilarity.py]: Good idea, thanks!

File: matchms/similarity/CosineGreedy.py [hunk 2]
```diff
@@ -34,7 +35,7 @@ class CosineGreedy:
         # Use factory to construct a similarity function
         cosine_greedy = CosineGreedy(tolerance=0.2)
 
-        score, n_matches = cosine_greedy(spectrum_1, spectrum_2)
+        score, n_matches = cosine_greedy.pair(spectrum_1, spectrum_2)
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply matchms/similarity/CosineGreedy.py]: True, done.

File: matchms/similarity/BaseSimilarityFunction.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+class BaseSimilarityFunction:
+    """Similarity function base class.
+    When building a custom similarity measure, inherit from this class and implement
+    the desired methods.
+    """
+    # Set key characteristics as class attributes
+    is_commutative = True
+
+    def pair(self, reference: SpectrumType, query: SpectrumType) -> float:
+        """Required: Method to calculate the similarity for one input pair.
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply matchms/similarity/BaseSimilarityFunction.py]: Looks nicer. Thanks.

File: matchms/similarity/__init__.py [hunk 4]
```diff
@@ -8,28 +8,26 @@
 * scores based on comparing peak positions and intensities
   (:meth:`~matchms.similarity.CosineGreedy` or :meth:`~matchms.similarity.ModifiedCosine`)
 * simple scores that only assess parent mass matches
-  (:meth:`~matchms.similarity.ParentmassMatchParallel`)
+  (:meth:`~matchms.similarity.ParentmassMatch`)
 * scores assessing molecular similarity if structures (SMILES, InchiKey) are given as metadata
-  (:meth:`~matchms.similarity.FingerprintSimilarityParallel`)
+  (:meth:`~matchms.similarity.FingerprintSimilarity`)
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply matchms/similarity/__init__.py]: Thanks. Done.

File: matchms/similarity/BaseSimilarityFunction.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+from matchms.typing import SpectrumType
+
+
+class BaseSimilarityFunction:
+    """Similarity function base class.
+    When building a custom similarity measure, inherit from this class and implement
+    the desired methods.
+    """
+    # Set key characteristics as class attributes
+    is_commutative = True
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply matchms/similarity/BaseSimilarityFunction.py]: Good point, thanks! Done.

File: matchms/similarity/BaseSimilarityFunction.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        is_symmetric
+            Set to True when *references* and *queries* are identical (as for instance for an all-vs-all
+            comparison). By using the fact that score[i,j] = score[j,i] the calculation will be about
+            2x faster.
+        """
+        n_rows = len(references)
+        n_cols = len(queries)
+        scores = numpy.empty([n_rows, n_cols], dtype="object")
+        for i_ref, reference in enumerate(references[:n_rows]):
+            if is_symmetric:
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply matchms/similarity/BaseSimilarityFunction.py]: Thanks! Done.

File: matchms/similarity/CosineGreedy.py [hunk 7]
```diff
@@ -14,9 +15,9 @@ class CosineGreedy:
     m/z ratios lie within the given 'tolerance'.
     The underlying peak assignment problem is here solved in a 'greedy' way.
     This can perform notably faster, but does occasionally deviate slightly from
-    a fully correct solution (as with the Hungarian algorithm). In practice this
-    will rarely affect similarity scores notably, in particular for smaller
-    tolerances.
+    a fully correct solution (as with the Hungarian algorithm, see
+    :meth:`~matchms.similarity.CosineHungarian`). In practice this will rarely
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply matchms/similarity/CosineGreedy.py]: Done

File: matchms/similarity/CosineHungarian.py [hunk 8]
```diff
@@ -4,19 +4,20 @@
...
     The cosine score aims at quantifying the similarity between two mass spectra.
     The score is calculated by finding best possible matches between peaks
     of two spectra. Two peaks are considered a potential match if their
     m/z ratios lie within the given 'tolerance'.
     The underlying peak assignment problem is here solved using the Hungarian algorithm.
-    This can perform notably slower than the 'greedy' implementation in CosineGreedy, but
-    does represent a mathematically proper solution to the problem.
-
+    This can perform notably slower than the 'greedy' implementation in
+    :meth:`~matchms.similarity.CosineGreedy`, but does represent a mathematically proper
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply matchms/similarity/CosineHungarian.py]: Done.

File: matchms/Scores.py [hunk 9]
```diff
@@ -60,15 +64,20 @@ def __init__(self, references: ReferencesType, queries: QueriesType, similarity_
...
-        Scores._validate_input_arguments(references, queries, similarity_function)
+        # pylint: disable=too-many-arguments
+        Scores._validate_input_arguments(references, queries)
 
         self.n_rows = len(references)
         self.n_cols = len(queries)
-        self.references = numpy.asarray(references).reshape(self.n_rows, 1)
-        self.queries = numpy.asarray(queries).reshape(1, self.n_cols)
+        self.references = numpy.asarray(references)
+        self.queries = numpy.asarray(queries)
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply matchms/Scores.py:80]: Thank you!

File: matchms/Scores.py [hunk 10]
```diff
@@ -60,15 +64,20 @@ def __init__(self, references: ReferencesType, queries: QueriesType, similarity_
             List of query objects
         similarity_function
             Function which accepts a reference + query object and returns a score or tuple of scores
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply matchms/Scores.py]: I changed the docstring and linked to ``BaseSimilarity``.

File: matchms/similarity/FingerprintSimilarity.py [hunk 11]
```diff
@@ -1,20 +1,64 @@
...
 
-class FingerprintSimilarityParallel:
+class FingerprintSimilarity(BaseSimilarityFunction):
     """Calculate similarity between molecules based on their fingerprints.
 
     For this similarity measure to work, fingerprints are expected to be derived
-    by running the "add_fingerprint()" filter function.
+    by running :meth:`~matchms.filtering.add_fingerprint`.
+
+    Code example:
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply matchms/similarity/FingerprintSimilarity.py:20]: Yay! :)

File: matchms/Scores.py [hunk 12]
```diff
@@ -83,42 +92,35 @@ def __next__(self):
...
     @staticmethod
-    def _validate_input_arguments(references, queries, similarity_function):
+    def _validate_input_arguments(references, queries):
         assert isinstance(references, (list, tuple, numpy.ndarray)),\
             "Expected input argument 'references' to be list or tuple or numpy.ndarray."
 
         assert isinstance(queries, (list, tuple, numpy.ndarray)),\
             "Expected input argument 'queries' to be list or tuple or numpy.ndarray."
 
-        assert callable(similarity_function), "Expected input argument 'similarity_function' to be callable."
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply matchms/Scores.py:101]: ~~That gives a pylint complaint:
``pylint: isinstance-second-argument-not-valid-type / Second argument of isinstance is not a type``~~
(sorry, wrong import of BaseSimilarity class...)

File: matchms/Scores.py [hunk 13]
```diff
@@ -60,15 +64,20 @@ def __init__(self, references: ReferencesType, queries: QueriesType, similarity_
...
         similarity_function
             Function which accepts a reference + query object and returns a score or tuple of scores
-
+        is_symmetric
+            Set to True when *references* and *queries* are identical (as for instance for an all-vs-all
+            comparison). By using the fact that score[i,j] = score[j,i] the calculation will be about
+            2x faster.
         """
-        Scores._validate_input_arguments(references, queries, similarity_function)
+        # pylint: disable=too-many-arguments
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply matchms/Scores.py:74]: Still complained when I removed it.

File: matchms/Scores.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (florian-huber) [reply matchms/Scores.py:101]: Done.

**Comment by florian-huber** [comment]
Thanks a lot for the review Stefan!
I now changed ``BaseSimilarityFunction`` to ``BaseSimilarity`` and added changes to the Changelog.

File: matchms/Scores.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (sverhoeven) [reply matchms/Scores.py:74]: ok

**Review by sverhoeven — CHANGES_REQUESTED** [review]
Almost there, whats there looks good.

However some files not-changed by PR still need updating:
* [calculate_scores()](https://github.com/matchms/matchms/blob/refactor_v2/matchms/calculate_scores.py#L9) 
  * still has old type for similarity function, it should be matchms.similarity.BaseSimilarity.BaseSimilarity
  *  method needs `is_symmetric` argument, so Scores.calculate can be deprecated without losing functionality.
* please remove SimilarityFunction from matchms/typing.py to avoid confusion

**Comment by florian-huber** [comment]
Thanks a lot @sverhoeven . I removed the old typing and added ``is_symmetric`` to ``calculate_scores()``.

**Review by sverhoeven — APPROVED** [review]
Everything is order, ready to merge

**Comment by florian-huber** [comment]
Thanks a lot Stefan to help with this beast :)

**Pull request merged**